### DISCUSSION TITLE
A general question about pretraining and training of dialogue systems

### DISCUSSION THREAD

**Discussion by helehh** [facebookresearch/ParlAI#4068] [Q&A] [answered]
Hi!

I am rather new to open-domain dialogue systems and while reading the papers (for example [Recipes for building an open-domain chatbot](https://arxiv.org/abs/2004.13637)) about the topic I noticed that there's a detail about the training that is never explained (not only in this paper but in many papers that I checked). More precisely, it is always said that pretraining of a generative model is done by _generating a comment conditioned on the full thread leading up to the comment_. However, it is never said whether one thread counts as only one training example or is it used to construct multiple (N-1) training examples.

To illustrate my question, let's say we have a following thread:
A: Hello!
B: Hi!
A: How are you?
B: Good. How are you?

Would it be a single training example:
1. Context: Hello! Hi! How are you?, Label: Good. How are you?

or is it used to create multiple examples:
1. Context: Hello!, Label: Hi!
2. Context: Hello! Hi!, Label: How are you? 
3. Context: Hello! Hi! How are you?, Label: Good. How are you?

In addition, is the same (either one) approach applied during fine-tuning on multi-turn dataset?

Thank you in advance!

**Comment by stephenroller** [comment, marked as answer]
We train to predict every non-root node in the tree.

```
    A
   / \
  B   C
 / \
E   F
```

Then we generate 4 (N-1) examples:

Context -> Label
A -> B
A -> C
A, B -> E
A, B -> F

**Reply by helehh** [reply]
Thank you for this clarification, Stephen!

**Reply by helehh** [reply]
Hi again, @stephenroller!

Am I correct that in terms of ParlAI Dialog Format each example would count as one episode, e.g. the training data for pre-training would look like:
```
text:A    label: B    episode_done:True
text:A    label: C    episode_done:True
text:A B    label: E    episode_done:True
text:A B    label: F    episode_done:True
```

**Reply by stephenroller** [reply]
Correct
